xrootd/xrdproto/protocol.go

```diff
@@ -322,26 +322,29 @@ func WriteResponse(w io.Writer, streamID StreamID, status ResponseStatus, resp M
 // ReadResponse returns the response header and the response body.
 // ReadResponse requires serialization since multiple ReadFull calls are made.
 func ReadResponse(r io.Reader) (ResponseHeader, []byte, error) {
-	var responseHdr ResponseHeader
-	headerBytes := make([]byte, ResponseHeaderLength)
+	var header ResponseHeader
+	data, err := ReadResponseWithReuse(r, make([]byte, ResponseHeaderLength), &header)
+	return header, data, err
+}
+
+// ReadResponseWithReuse reads a XRootD response from r. A response header is read into headerBytes and
+// unmarshaled to header for the reusing reasons.
+// ReadResponseWithReuse returns the response body.
+// ReadResponseWithReuse requires serialization since multiple ReadFull calls are made.
+func ReadResponseWithReuse(r io.Reader, headerBytes []byte, header *ResponseHeader) ([]byte, error) {
```

Comment:
it seems to me (but I may be mistaken) we could completely drop `ReadResponse` and replace it with this `ReadResponseWithReuse` (that we could then rename into `ReadResponse`).

we could also slightly modify its signature:
```go
func ReadResponse(r io.Reader, hdr *ResponseHeader) ([]byte, error) {
    var buf [ResponseHeaderLength]byte
    if _, err := io.ReadFull(r, buf[:]); err != nil { ... }
    ...
}
```

what do you think?

Reply:
That adds an extra allocation of  for each `ReadResponse` call. 
`+8 B/op, +1 allocs/op` for `BenchmarkSession_Ping`.

To be honest, I'm not sure if that's so important: we are losing a bit in readability and acquire a bit in performance, or vice-versa.

I'll be happy to hear your opinion! :)

Reply:
nah... this can be tabled for later (if really needed.)
it was just a late night idea :)